### DISCUSSION TITLE
Convenient way of rendering remote images

### DISCUSSION THREAD

**Discussion by gyzerok** [emilk/egui#1669] [Q&A]
Hello!

I am using `egui` to write a client messenger application. One of the things I struggle with a lot - is rendering remote images such as avatars, attachments and such. In order to do it I am currently loading them (in background thread) and storing in application state.

There are 2 major inconveniences with it (from my perspective):
1. I need to go through all my models when I load them and load all the images from them into application state. Since I am doing it by hand I have to remember to do it for every single field which contains image URL
2. I then need to pass all this state around in order to render those images

From my perspective doing it for images which are static resources is good enough, but for my use case it is a struggle.

What I would love to have is some custom widget that upon rendering would check if image was already loaded somewhere (perhaps `ctx.memory()`?) and then load it if not (perhaps in another thread).

```rust
RemoteImage::new(user.avatar_url()).show(ui)
```

However I am not experienced enough in Rust and egui to be able to implement this widget myself.

I do understand that creating such widget as part of egui probably won't work since different applications might have different background threads setups. However maybe it would be possible to create a reference implementation in examples for such a use case?

**Comment by emilk** [comment]
Adding a image/texture manager that dynamically loads images when needed, and frees them when no longer needed, would be great. There is some discussion of it in https://github.com/emilk/egui/discussions/1513

**Comment by gyzerok** [comment]
Thank you - some great discussions there. I could definitely use some of it.

One question though is still unanswered by mentioned issue. How would I store my custom texture manager inside context, so I don't need to pass it around everywhere to all of my widgets? It looks like `Context::data()` and `Context::memory` are there for other purposes.

I think this is currently my biggest problem. If this one is solved, then everything else is easy :)

**Reply by emilk** [reply]
You could maybe use `Context::data`, but really `egui::Context` is not meant to be a place for "store extra parameters to your function here" :)
